src/stochastic_swap.rs

```diff
@@ -0,0 +1,314 @@
+/// the thread pool to 4 threads.
+///
+/// Args:
+///     num_trials (int): The number of random trials to attempt
+///     num_qubits (int): The number of qubits
+///     int_layout (NLayout): The initial layout for the layer. The layout is a mapping
+///         of virtual qubits to physical qubits in the coupling graph
+///     int_qubit_subset (ndarray): A 1D array of qubit indices for the set of qubits in the
+///         coupling map that we've chosen to map into.
+///     int_gates (ndarray): A 1D array of qubit pairs that each 2 qubit gate operates on.
+///         The pairs are flattened on the array so that each pair in the list of 2q gates
+///         are adjacent in the array. For example, if the 2q interaction list was
+///         ``[(0, 1), (2, 1), (3, 2)]``, the input here would be ``[0, 1, 2, 1, 3, 2]``.
+///     cdist (ndarray): The distance matrix for the coupling graph of the target
+///         backend
+///     cdist2 (ndarray): The distance matrix squared for the coupling graph of the
+///         target backend
+///     edges (ndarray): A flattened 1d array of the edge list of the coupling graph.
+///         The pairs are flattened on the array so that each node pair in the edge are
+///         adjacent in the array. For example, if the edge list were ``[(0, 1), (1, 2), (2, 3)]``
+///         the input array here would be ``[0, 1, 1, 2, 2, 3]``.
+///     seed (int): An optional seed for the rng used to generate the random perturbation
+///         matrix used in each trial
+/// Returns:
+///     tuple: If a valid layout permutation is found a tuple of the form:
+///         ``(edges, layout, depth)`` is returned. If a solution is not found the output
+///         will be ``(None, None, max int)``.
+#[pyfunction]
+#[pyo3(
+    text_signature = "(num_trials, num_qubits, int_layout, int_gates, cdist, cdist2, edges, /, seed)"
+)]
+pub fn swap_trials(
+    num_trials: u64,
+    num_qubits: usize,
+    int_layout: &NLayout,
+    int_qubit_subset: PyReadonlyArray1<usize>,
+    int_gates: PyReadonlyArray1<usize>,
+    cdist: PyReadonlyArray2<f64>,
+    cdist2: PyReadonlyArray2<f64>,
+    edges: PyReadonlyArray1<usize>,
+    seed: Option<u64>,
+) -> PyResult<(Option<EdgeCollection>, Option<NLayout>, usize)> {
+    let int_qubit_subset_arr = int_qubit_subset.as_slice()?;
+    let int_gates_arr = int_gates.as_slice()?;
+    let cdist_arr = cdist.as_array();
+    let cdist2_arr = cdist2.as_array();
+    let edges_arr = edges.as_slice()?;
+    let num_gates: usize = int_gates.len() / 2;
+    let mut best_possible: Option<(u64, f64, EdgeCollection, NLayout)> = None;
+    let locked_best_possible: RwLock<&mut Option<(u64, f64, EdgeCollection, NLayout)>> =
+        RwLock::new(&mut best_possible);
+    let mut outer_rng: Pcg64Mcg = match seed {
+        Some(seed) => Pcg64Mcg::seed_from_u64(seed),
+        None => Pcg64Mcg::from_entropy(),
+    };
+    let seed_vec: Vec<u64> = vec![outer_rng.gen(); num_trials as usize];
```

Comment:
I believe the expression `outer_rng.gen()` will be evaluated once and this means all elements of `seed_vec` will be equal.

```suggestion
    let seed_vec: Vec<u64> = outer_rng
        .sample_iter(&rand::distributions::Standard)
        .take(n)
        .collect();
```

Reply:
Oops, yep good catch. Heh I think this mistake is actually how I fixed the race condition above too, made it deterministic by using the same seed in each thread local rng. Trying this locally it looks like there is now non-deterministic behavior again :)